pkg/storegateway/bucket_stores.go

```diff
@@ -114,8 +114,9 @@ func NewBucketStores(cfg tsdb.BlocksStorageConfig, shardingStrategy ShardingStra
 	lazyLoadingGate := gate.NewNoop()
 	lazyLoadingMax := cfg.BucketStore.IndexHeader.LazyLoadingConcurrency
 	if lazyLoadingMax != 0 {
-		blockingGate := gate.NewBlocking(cfg.BucketStore.IndexHeader.LazyLoadingConcurrency)
-		lazyLoadingGate = gate.NewInstrumented(lazyLoadingGateReg, cfg.BucketStore.IndexHeader.LazyLoadingConcurrency, blockingGate)
+		lazyLoadingGate = gate.NewBlocking(cfg.BucketStore.IndexHeader.LazyLoadingConcurrency)
+		lazyLoadingGate = gate.NewInstrumented(lazyLoadingGateReg, cfg.BucketStore.IndexHeader.LazyLoadingConcurrency, lazyLoadingGate)
+		lazyLoadingGate = timeoutGate{delegate: lazyLoadingGate, timeout: cfg.BucketStore.IndexHeader.LazyLoadingConcurrencyQueueTimeout}
```

Comment:
should we cancel the lazy loading or keep it running in the background? I feel like we'd be throwing away work if we cancel it. Given we just received a query for a block, same block may be likely queried again soon

Reply:
If it gets cancelled from hitting the timeout it hadn't started loading anything yet (since the max number of index-headers were already being loaded, this one had to wait) and so we aren't throwing away any work for that particular block. We'd be throwing away other work done for the query but not index-headers being loaded. Does that make sense?

Reply:
Oh, I see what you mean. Keep the load running so it gets loaded eventually even after we timeout the request and return an error. I guess we could but I'd rather keep the first iteration of this simple so we can get it merged and see how it behaves before doing anything more complex.

Reply:
I'm actually wrong. The timeout applies only on the queuing part, not on the loading. So even if loading takes longer that the timeout of the imeout gate, it won't cancel the request. This looks good.